src/main/java/com/avispl/symphony/dal/avdevices/audionetworkinterface/qsc/undnemo/QSCUndnemoCommunicator.java

```diff
@@ -0,0 +1,805 @@
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import org.springframework.util.CollectionUtils;
+
+import com.avispl.symphony.api.dal.control.Controller;
+import com.avispl.symphony.api.dal.dto.control.AdvancedControllableProperty;
+import com.avispl.symphony.api.dal.dto.control.ControllableProperty;
+import com.avispl.symphony.api.dal.dto.monitor.ExtendedStatistics;
+import com.avispl.symphony.api.dal.dto.monitor.Statistics;
+import com.avispl.symphony.api.dal.error.CommandFailureException;
+import com.avispl.symphony.api.dal.error.ResourceNotReachableException;
+import com.avispl.symphony.api.dal.monitor.Monitorable;
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.dto.ChannelInfo;
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.utils.QSCUndnemoConstant;
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.utils.QSCUndnemoMetric;
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.utils.QSCUndnemoUDPCommand;
+import com.avispl.symphony.dal.util.StringUtils;
+
+/**
+ * QSC Attero Tech unDNEMO Adapter
+ *
+ * Monitoring:
+ * <ul>
+ * 	<li>Online / Offline Status</li>
+ * 	<li>Version Info</li>
+ * 	<li>Channel Info (channels 1-64)</li>
+ * 	<li>Active Channel Index (1-64)</li>
+ * 	<li>Speaker Mute (On/Off)</li>
+ * 	<li>Volume (1-10)</li>
+ * 	<li>Button Brightness (0-10)</li>
+ * 	<li>Display Brightness (0-10)</li>
+ * </ul>
+ *
+ * Controlling:
+ * <ul>
+ * 	<li>Set Active Channel Index (1-64)</li>
+ * 	<li>Set Speaker Mute (On/Off)</li>
+ * 	<li>Set Volume (1-10)</li>
+ * 	<li>Set Button Brightness (0-10)</li>
+ * 	<li>Set Display Brightness (0-10)</li>
+ * 	</ul
+ *
+ * @author Duy Nguyen
+ * @version 1.0
+ * @since 1.0
+ */
+public class QSCUndnemoCommunicator extends UDPCommunicator implements Monitorable, Controller {
+
+	/**
+	 * Process that is running whenever {@link QSCUndnemoCommunicator#getMultipleStatistics()} is called to fetch all 64 channel information.
+	 * - The worker thread will be destroyed when all 64 channels are fetched successfully or {@link QSCUndnemoCommunicator#internalDestroy()} is called.
+	 *
+	 * @author Maksym.Rossiytsev, Duy Nguyen
+	 * @since 1.0.0
+	 */
+	class QSCChannelDataLoader implements Runnable {
+
+		private volatile boolean inProgress;
+
+		private List<Integer> listIndexes;
+
+		/**
+		 * QSCChannelDataLoader with args constructor
+		 *
+		 * @param indexes list of index
+		 */
+		public QSCChannelDataLoader(List<Integer> indexes) {
+			inProgress = true;
+			listIndexes = indexes;
+		}
+
+		@Override
+		public void run() {
+			mainloop:
+			while (inProgress) {
```

Comment:
I think you should remove this "while", It's keeping the thread run forever. For every interval, you submit the new thread. It will make the memory overflow

Reply:
You're right. Updated as well.